Adopt dynamicDowncast<> in MarkupAccumulator

annevk opened the pull request:
#### d1d58f130250d33153da12f2ae860715beaeb7dd
<pre>
Adopt dynamicDowncast&lt;&gt; in MarkupAccumulator
<a href="https://bugs.webkit.org/show_bug.cgi?id=268220">https://bugs.webkit.org/show_bug.cgi?id=268220</a>

Reviewed by Chris Dumez.

* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::elementCannotHaveEndTag):
(WebCore::MarkupAccumulator::appendContentsForNode):
(WebCore::MarkupAccumulator::serializeNodesWithNamespaces):
(WebCore::MarkupAccumulator::resolveURLIfNeeded const):
(WebCore::MarkupAccumulator::replacementElement):
(WebCore::MarkupAccumulator::startAppendingNode):
(WebCore::MarkupAccumulator::frameForAttributeReplacement const):

Canonical link: <a href="https://commits.webkit.org/273621@main">https://commits.webkit.org/273621@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/988e5e07b03b1038c130a9da8f43b907c193f44d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36120 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15067 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38322 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38841 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32472 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37342 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17505 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12095 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31168 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36675 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12725 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32044 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11154 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11176 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32220 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40087 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32790 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32584 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37115 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11370 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9260 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35200 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13095 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8208 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11849 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12208 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -336,26 +338,26 @@ String MarkupAccumulator::resolveURLIfNeeded(const Element& element, const Strin
 {
-    if (!m_shouldIncludeShadowDOM || !node.isShadowRoot())
+    RefPtr shadowRoot = dynamicDowncast<ShadowRoot>(node);
+    if (!shadowRoot || !m_shouldIncludeShadowDOM)
```
Might be nice to do the `!m_shouldIncludeShadowDOM` early return earlier, in order to avoid the dynamicDowncast<>() in some cases.

cdumez reviewed (approved):
LGTM

annevk committed:
Adopt dynamicDowncast<> in MarkupAccumulator
https://bugs.webkit.org/show_bug.cgi?id=268220

Reviewed by Chris Dumez.

* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::elementCannotHaveEndTag):
(WebCore::MarkupAccumulator::appendContentsForNode):
(WebCore::MarkupAccumulator::serializeNodesWithNamespaces):
(WebCore::MarkupAccumulator::resolveURLIfNeeded const):
(WebCore::MarkupAccumulator::replacementElement):
(WebCore::MarkupAccumulator::startAppendingNode):
(WebCore::MarkupAccumulator::frameForAttributeReplacement const):

Canonical link: https://commits.webkit.org/273621@main
